torch/_dynamo/backends/train_step.py

```diff
@@ -0,0 +1,181 @@
+    orig_states: Dict[str, Any] = copy(opt.state)
+    for n in named_states:
+        # opt.state's key type is string, but optimizer uses Parameter as keys
+        opt.state[params[n]] = named_states[n]  # type: ignore[index]
+
+    # FIXME: support multiple parameter groups
+    param_group = opt.param_groups[0]
+    orig_params = param_group["params"]
+    # FIXME(@mrshenli): exclude buffers
+    print(f"setting optimizer to use parm0 {params}")
+    param_group["params"] = params.values()
+
+    try:
+        yield
+    finally:
+        param_group["params"] = orig_params
+        opt.state.update(orig_states)
+
+
+def train_step_compiler(backend_compile_fn):
+    """Note [Train Step Compile]
+
+    Usually, torch.compile() allows graph-breaks and compiles pairs of forward (+backward) by
+    extracting sections of forward from python programs and using AotAutograd to produce corresponding
+    chunks of backwards, tying it back together with an AotFunction.
+
+    Instead, TrainStepCompiler assumes the user compiles a full train_step function complete with calls to
+    .backward(), optimizer step(), and zero_grad().  It additionally requires no graph-breaks.
+
+    Args:
+        backend_compile_fn (callable): A dynamo compiler function, to be invoked to compile each subgraph.
+    """
+
+    def _compile_fn(mod: fx.GraphModule, fake_inputs: List[torch.Tensor]):
+        """
+        Step 1: Assert inputs (from user) are already Fake, and user their FakeTensorMode
+                (created by dynamo) to fakeify the module's parameters
+        """
+        # at input dynamo puts two copies e.g.
+        #   L__model___layers_0.weight
+        #   model.layers.0.weight
+        # both as realtensor on passed in graphmodule.  `model` comes from having model as input to train_step.
+
+        print(mod.graph)
+        torch.set_grad_enabled(True)
+        torch._dynamo.utils.assert_no_fake_params_or_buffers(mod)
+        assert len(fake_inputs) > 0, "Expected at least one input"
+        fake_mode = fake_inputs[0].fake_mode
+        assert isinstance(
+            fake_mode, FakeTensorMode
+        ), "Expected a valid FakeTensorMode on dynamo inputs"
+
+        def fakeify_inputs(flat_args):
+            already_fake = {}
+            def convert(idx, x):
+                # todo: do we expect symint inputs?
+                assert isinstance(x, torch.Tensor)
+                if x not in already_fake:
+                    # Since we do have duplicate names from dynamo refering to the same tensor,
+                    # ensure that we never make more than one faketensor for a given real tensor!
+                    already_fake[x] = fake_mode.from_tensor(x, static_shapes=False)
+                return already_fake[x]
+
+            return [convert(idx, x) for idx, x in enumerate(flat_args)]
+
+        # problem: if i don't remove duplicates here, stateless.Parametrize gets confused by presence of duplicates
```

Comment:
@albanD this is something strange i am working around for the moment but wanted some help to do properly.

Basically, dynamo puts 2 names to the same tensor on my graphmodule:
  mod.model.layers[0].weight
  and
  mod.__layers_0_weight (or something like that..)


the `__layers_0_weight` name is the one made by dynamo when tracing a usage of the param, so in my simple case that means all uses of the weight are used via this name, and I am getting away with only fake-ifying this name and leaving `model.*` as real tensors when I do the fx tracing.  It might be the case that dynamo can fully gaurantee this, but i'm not sure if it's globally safe or not.

I wanted to fakify all names instead, but the issue I ran into is that 'reparametrize_module' would never do the equivalent of `remove_duplicate=True` for me and the accessor.swap was returning an 'orig_module_state' that had real tensors for the names like '__layers_0_weight' but had my provided faketensors (provided to reparametrize) for names like `model.layers[0].weight`.  Then after the yield, parametrize would 'restore' these faketensors to replace the actual real tensors on the module.

Reply:
I'm not sure to follow exactly. Duplicated params are properly supported in stateless.functioncal_call(). 
If you're using private functions in there I'm not sure though.